diff --git a/dix/getevents.c b/dix/getevents.c
index b0cb3bc64..84abf80e8 100644
--- a/dix/getevents.c
+++ b/dix/getevents.c
@@ -1262,8 +1262,26 @@ storeLastValuators(DeviceIntPtr dev, ValuatorMask *mask, double devx, double dev
             dev->last.valuators[i] = valuator_mask_get_double(mask, i);
     }
 
+
 }
 
+void
+QueuePointerRelativeMotion2D(DeviceIntPtr device, double dx, double dy,
+                             double raw_dx, double raw_dy)
+{
+    int nevents;
+    ValuatorMask mask = {0};
+
+    valuator_mask_set_unaccelerated(&mask, 0, dx, raw_dx);
+    valuator_mask_set_unaccelerated(&mask, 1, dy, raw_dy);
+
+    nevents = GetPointerEvents(InputEventList, device, MotionNotify, 0,
+                               POINTER_RELATIVE, &mask);
+
+    queueEventList(device, InputEventList, nevents);
+}
+
+
 /**
  * Generate internal events representing this pointer event and enqueue them
  * on the event queue.
diff --git a/include/input.h b/include/input.h
index de89ee37d..0f4a43002 100644
--- a/include/input.h
+++ b/include/input.h
@@ -437,6 +437,10 @@ extern _X_EXPORT void QueuePointerEvents(DeviceIntPtr pDev,
                                          int buttons,
                                          int flags, const ValuatorMask *mask);
 
+extern _X_EXPORT void QueuePointerRelativeMotion2D(DeviceIntPtr pDev,
+                                                   double dx, double dy,
+                                                   double raw_dx, double raw_dy);
+
 extern _X_EXPORT int GetKeyboardEvents(InternalEvent *events,
                                        DeviceIntPtr pDev,
                                        int type,
